File: ruby/tree_traversal_inorder.rb
class Node
  attr_accessor :value, :left, :right

  def initialize(value, left = nil, right = nil)
    @value = value
    @left = left
    @right = right
  end
end

def tree_traversal_inorder(root)
 retrun if root.nil?

 tree_traversal_inorder(root.left) #{ Traverse left subtree}
 puts root.value #Visit root
 tree_traversal_inorder(root.rigt) #{ Traverse right subtree}
end

if __FILE__ == $PROGRAM_NAME
  root = Node.new(2, Node.new(-10), Node.new(20))
  puts "Expecting: [-10, 2, 20]"
  print tree_traversal_inorder(root)
  puts

  puts

  root = Node.new(10, Node.new(0, nil, Node.new(5)), Node.new(20, nil, Node.new(30)))
  puts "Expecting: [0, 5, 10, 20, 30] "
  print tree_traversal_inorder(root)
  puts

  root = Node.new(30, Node.new(10, nil, Node.new(20, nil, Node.new(25, Node.new(24)))), Node.new(50, Node.new(40, Node.new(39)), Node.new(55, Node.new(54), Node.new(56, nil, Node.new(60)))))
  puts "Expecting: [5, 6, 7, 8, 9, 10] "
  print tree_traversal_inorder(root)
  puts

root = Node.new(2, Node.new(-10), Node.new(20))
puts "Expecting: [-10, 2, 20]"
tree_traversal_inorder(root)
puts ""

root = Node.new(10, Node.new(0, nil, Node.new(5)), Node.new(20, nil, Node.new(30)))
puts "Expecting: [0, 5, 10, 20, 30]"
tree_traversal_inorder(root)
puts ""





  #{}Don't forget to add your own!
end

# Please add your pseudocode to this file : Pseudocode
#{ class Node
#{Attr_accessor : value, : left, :right = nil
#@value = value
#@left = left
#@right = right
  #end
#end

#{def tree_traversal_border(root)
#return if root is nil


#{trrr-traversal_inorder(root.left) // Traverse left subteet
#{print root.value//Visit root
#{tree_traversal_inorder{root.right) //Traverse right subtree}
#{tree_traversal_inorder(root.rigt) // Traverse right subtree}
#end

#{def intialize(value, left = nil, right ,)}
# And a written explanation of your solution
#{#{#{#{}}}}

File: javascript/tree_traversal_inorder.js
class Node {
  constructor(value, left = null, right = null) {
    this.value = value;
    this.left = left;
    this.right = right;
  }
}

function treeTraversalInorder(root) {
  if (root === null) {
    return;
  }

  treeTraversalInorder(root.left); // Traverse left subtree
  console.log(root.value); // Visit root
  treeTraversalInorder(root.right); // Traverse right subtree
}

if (require.main === module) {
  // add your own tests in here

  let root = new Node(5);
  console.log("Expecting: [5]");
  treeTraversalInorder(root);
  console.log("");

  root = new Node(10, new Node(5), new Node(15));
  console.log("Expecting: [5, 10, 15]");
  treeTraversalInorder(root);
  console.log("");

  root = new Node(50, new Node(30, new Node(20), new Node(40)), new Node(70, new Node(60), new Node(80)));
  console.log("Expecting: [20, 30, 40, 50, 60, 70, 80]");
  treeTraversalInorder(root);
  console.log("");
}

module.exports = {
  Node,
  treeTraversalInorder
};

  /*Pseudocode;

/*function treeTraversalInorder(root):
if root is null:
return

//treeTraversalInorder(root.left) // Traverse left subtree//print root.value // Visit root
//treeTraversalInorder(root.right) // Traverse right subtree


Explanation:
// The Node class represents a node in the binary tree, with a value and left and right child nodes.//The treeTraversalInorder function performs the inorder traversal of the binary tree recursively.
// In the treeTraversalInorder function, we first check if the root is null. If it is, we return from the function.
// Then, we recursively call treeTraversalInorder on the left subtree, visit the root node by printing its value, and finally, recursively call treeTraversalInorder on the right subtree.
// The test cases provided create two different binary trees and call the treeTraversalInorder function on them to verify the correctness of the implementation.
// The expected output is [-10, 2, 20] for the first tree and [0, 5, 10, 20, 30] for the second tree.
*/